### PR TITLE
ODS-5568 - Create separate parameters for creating apisdk and testsdk packages

### PR DESCRIPTION
**Pull request by bptillman**
This resolved the reported issue since what was happening was the testsdk and apisdk were not being generated correctly 
 in the right builds, so the testsdk that was generated was just the renamed apisdk that did not contain all the necessary resources in the contained dll. The approach to fix this was to introduce two new parameters for handling the package generation, since we need to be able to generate one or the other, or both. These two new parameters control the logic flow that would allow you to generate them in that same combination.

The new parameters referenced in this PR have been added to the initdev template build [here](https://intedfitools1.msdf.org/admin/editBuildParams.html?id=template:OdsPlatform_OdsApiInitDevUnitTestPackageTemplate) , with both set to false, matching the previous parameter that was also set to false. This template is used for initdev unit test, initdev integration test, and initdev postgresql. On the [TestSdk](https://intedfitools1.msdf.org/admin/editBuildParams.html?id=buildType:OdsPlatform_OdsApi_Packages_EdFiOdsApiTestSdk) build, the parameters were also aded and only generating the TestSdk is set to true. The ApiSdk build was not modified since it does not use the initdev script to build its package.

image

Once this PR is merged the parameter `odsapi.build.generateSdkPackages` will be removed from all builds configurations since it will no longer exist.

### CODE REVIEW

File: .github/workflows/InitDev (Postgres),Unit tests,Integration tests.yml [hunk 1]
```diff
@@ -90,7 +90,7 @@ jobs:
         [int]$BuildIncrementer = "${{ env.BUILD_INCREMENTER }}"
         [int]$newRevision =  $BuildCounter + $BuildIncrementer
         [string]$version = "${{ env.VERSION_MAJOR}}"+ "." +"${{ env.VERSION_MINOR}}" + "." + $newRevision.ToString()
-        .\build.github.ps1 -Engine PostgreSQL -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateSdkPackages -RunPostman -RunPester -RunDotnetTest -NoPackaging
+        .\build.github.ps1 -Engine PostgreSQL -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateApiSdkPackage -GenerateTestSdkPackage -RunPostman -RunPester -RunDotnetTest -NoPackaging
```
Comment (semalaiappan) [inline .github/workflows/InitDev (Postgres),Unit tests,Integration tests.yml:93]: Existing teamcity build is running with GenerateApiSdkPackage  only  . Is there any reason to include also GenerateApiSdkPackage ?   I am referring this build https://intedfitools1.msdf.org/buildConfiguration/OdsPlatform_OdsApi_OdsApiInitDevIntegrationTest?branch=%3Cdefault%3E&buildTypeTab=overview&mode=builds

Comment (semalaiappan) [inline .github/workflows/InitDev (Postgres),Unit tests,Integration tests.yml:93]: Existing teamcity build is running with GenerateApiSdkPackage  only  . Is there any reason to include also GenerateApiSdkPackage ?   I am referring this build https://intedfitools1.msdf.org/buildConfiguration/OdsPlatform_OdsApi_OdsApiInitDevIntegrationTest?branch=%3Cdefault%3E&buildTypeTab=overview&mode=builds

File: .github/workflows/InitDev,Unit tests,Integration tests,Package.yml [hunk 2]
```diff
@@ -111,7 +111,7 @@ jobs:
         [int]$BuildIncrementer = "${{ env.BUILD_INCREMENTER }}"
         [int]$newRevision =  $BuildCounter + $BuildIncrementer
         [string]$version = "${{ env.VERSION_MAJOR}}"+ "." +"${{ env.VERSION_MINOR}}" + "." + $newRevision.ToString()
-        .\build.github.ps1 -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateSdkPackages -RunPostman -RunPester -RunDotnetTest -PackageOutput "$env:GITHUB_WORKSPACE/Ed-Fi-ODS-Implementation/NugetPackages"
+        .\build.github.ps1 -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateApiSdkPackage -GenerateTestSdkPackage -RunPostman -RunPester -RunDotnetTest -PackageOutput "$env:GITHUB_WORKSPACE/Ed-Fi-ODS-Implementation/NugetPackages"
```
Comment (semalaiappan) [inline .github/workflows/InitDev,Unit tests,Integration tests,Package.yml:114]: Existing teamcity build is running with GenerateApiSdkPackage  only  . Is there any reason to include also GenerateApiSdkPackage ?   I am referring this build https://intedfitools1.msdf.org/buildConfiguration/OdsPlatform_OdsApi_OdsApiInitDevIntegrationTest?branch=%3Cdefault%3E&buildTypeTab=overview&mode=builds

File: logistics/scripts/Invoke-SdkGen.ps1 [hunk 3]
```diff
@@ -74,29 +76,36 @@ function Invoke-SdkGen {
...
+            $script:result += Invoke-Task "Restore-TestSdk-Packages" { Invoke-Restore-ApiSdk-Packages $sdkSolutionFile }
+            $script:result += Invoke-Task "Invoke-RebuildSolution" { Invoke-RebuildSolution $buildConfiguration "minimal" $sdkSolutionFile }
+
+            $nuspecFile = (Get-RepositoryResolvedPath "Utilities/SdkGen/EdFi.SdkGen.Console/EdFi.OdsApi.Sdk.nuspec")
+            $suffix = Get-ValueOrDefault $teamCityParameters['odsapi.package.suffix'] ".Suite3"
+            Write-Host "Updating package id and title to EdFi$suffix.OdsApi.TestSdk"
+            (Get-Content -path $nuspecFile -Raw) | ForEach-Object {
+                $_.replace("<id>EdFi.OdsApi.Sdk</id>","<id>EdFi$suffix.OdsApi.TestSdk</id>").replace("<title>EdFi.OdsApi.Sdk</title>","<title>EdFi$suffix.OdsApi.TestSdk</title>")
+            } | Set-Content -Path $nuspecFile
+            (Get-Content -path $nuspecFile -Raw) | ForEach-Object {
```
Comment (semalaiappan) [inline logistics/scripts/Invoke-SdkGen.ps1:104]: This small piece code is repeated as above 3 lines

**Review by semalaiappan — CHANGES_REQUESTED** [review]

**Comment by semalaiappan** [comment]
@bptillman  Please remove odsapi.build.generateSdkPackages variable from teamcity builds

File: .github/workflows/InitDev (Postgres), Tests using Sandbox, Year-Specific and Shared Instance.yml [hunk 4]
```diff
@@ -92,7 +92,7 @@ jobs:
         [int]$BuildIncrementer = "${{ env.BUILD_INCREMENTER }}"
         [int]$newRevision =  $BuildCounter + $BuildIncrementer
         [string]$version = "${{ env.VERSION_MAJOR}}"+ "." +"${{ env.VERSION_MINOR}}" + "." + $newRevision.ToString()
-        .\build.github.ps1 -Engine PostgreSQL -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateSdkPackages -NoPackaging
+        .\build.github.ps1 -Engine PostgreSQL -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateApiSdkPackage -GenerateTestSdkPackage -NoPackaging
```
Comment (semalaiappan) [inline .github/workflows/InitDev (Postgres), Tests using Sandbox, Year-Specific and Shared Instance.yml:95]: Why do we need to pass GenerateApiSdkPackage also . here we only need TEST SDK package correct ?

File: .github/workflows/InitDev,Unit tests,Integration tests,Smoke Tests using Sandbox, Year-Specific and Shared Instance.yml [hunk 5]
```diff
@@ -113,7 +113,7 @@ jobs:
         [int]$BuildIncrementer = "${{ env.BUILD_INCREMENTER }}"
         [int]$newRevision =  $BuildCounter + $BuildIncrementer
         [string]$version = "${{ env.VERSION_MAJOR}}"+ "." +"${{ env.VERSION_MINOR}}" + "." + $newRevision.ToString()
-        .\build.github.ps1 -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateSdkPackages -NoPackaging
+        .\build.github.ps1 -PackageVersion $version -RepositoryRoot $env:GITHUB_WORKSPACE -UsePlugins -RunSdkGen -GenerateApiSdkPackage -GenerateTestSdkPackage -NoPackaging
```
Comment (semalaiappan) [inline .github/workflows/InitDev,Unit tests,Integration tests,Smoke Tests using Sandbox, Year-Specific and Shared Instance.yml:116]: Why do we need to pass GenerateApiSdkPackage also . here we only need TEST SDK package correct ?

File: .github/workflows/InitDev (Postgres), Tests using Sandbox, Year-Specific and Shared Instance.yml (hunk 4 above)
Comment (bptillman) [reply .github/workflows/InitDev (Postgres), Tests using Sandbox, Year-Specific and Shared Instance.yml:95]: You are correct. I will remove GenerateApiSdkPackage.

File: .github/workflows/InitDev,Unit tests,Integration tests,Smoke Tests using Sandbox, Year-Specific and Shared Instance.yml (hunk 5 above)
Comment (bptillman) [reply .github/workflows/InitDev,Unit tests,Integration tests,Smoke Tests using Sandbox, Year-Specific and Shared Instance.yml:116]: You are correct. I will remove GenerateApiSdkPackage.

File: .github/workflows/InitDev,Unit tests,Integration tests,Package.yml (hunk 2 above)
Comment (bptillman) [reply .github/workflows/InitDev,Unit tests,Integration tests,Package.yml:114]: Per [ODS-4931](https://tracker.ed-fi.org/browse/ODS-4931), we should be generating both apisdk and testsdk as part of the initdev build.

File: .github/workflows/InitDev (Postgres),Unit tests,Integration tests.yml (hunk 1 above)
Comment (bptillman) [reply .github/workflows/InitDev (Postgres),Unit tests,Integration tests.yml:93]: Per [ODS-4931](https://tracker.ed-fi.org/browse/ODS-4931), we should be generating both apisdk and testsdk as part of the initdev build.

File: logistics/scripts/Invoke-SdkGen.ps1 (hunk 3 above)
Comment (bptillman) [reply logistics/scripts/Invoke-SdkGen.ps1:104]: It is repeated twice to cover both instances of how the nuspec file contents could need to be changed. The two paths covered are:
1. If we are only generating the test package, then the nuspec file would not have been modified yet and the id and title inside the nuspec will still be what is checked into source control, which is EdFi.OdsApi.Sdk. So we replace that with EdFi.Suite3.OdsApi.TestSdk
2. If we generate both api sdk package and test sdk package, then the nuspec file would have been modified for the api sdk package to have an id and title of EdFi.Suite3.OdsApi.Sdk, so we need to rename that to EdFi.Suite3.OdsApi.TestSdk

but looking at the code now. I think the two passes could be combined to something like this, where it does the 4 replaces together:
```
            (Get-Content -path $nuspecFile -Raw) | ForEach-Object {
                $_.replace("<id>EdFi.OdsApi.Sdk</id>","<id>EdFi$suffix.OdsApi.TestSdk</id>").replace("<title>EdFi.OdsApi.Sdk</title>","<title>EdFi$suffix.OdsApi.TestSdk</title>").replace("<id>EdFi$suffix.OdsApi.Sdk</id>","<id>EdFi$suffix.OdsApi.TestSdk</id>").replace("<title>EdFi$suffix.OdsApi.Sdk</title>","<title>EdFi$suffix.OdsApi.TestSdk</title>")
            } | Set-Content -Path $nuspecFile
```

File: .github/workflows/InitDev,Unit tests,Integration tests,Smoke Tests using Sandbox, Year-Specific and Shared Instance.yml (hunk 5 above)
Comment (bptillman) [reply .github/workflows/InitDev,Unit tests,Integration tests,Smoke Tests using Sandbox, Year-Specific and Shared Instance.yml]: This has now been removed

File: .github/workflows/InitDev (Postgres), Tests using Sandbox, Year-Specific and Shared Instance.yml (hunk 4 above)
Comment (bptillman) [reply .github/workflows/InitDev (Postgres), Tests using Sandbox, Year-Specific and Shared Instance.yml]: This has now been removed.

File: logistics/scripts/Invoke-SdkGen.ps1 (hunk 3 above)
Comment (bptillman) [reply logistics/scripts/Invoke-SdkGen.ps1]: I have pushed up a change to do the four replace calls in one line with one loop now.

File: .github/workflows/InitDev (Postgres),Unit tests,Integration tests.yml (hunk 1 above)
Comment (semalaiappan) [reply .github/workflows/InitDev (Postgres),Unit tests,Integration tests.yml:93]: OK Thanks

File: .github/workflows/InitDev,Unit tests,Integration tests,Package.yml (hunk 2 above)
Comment (semalaiappan) [reply .github/workflows/InitDev,Unit tests,Integration tests,Package.yml:114]: OK  Thanks

**Review by semalaiappan — APPROVED** [review]

**Pull request merged**